baseplate/core.py

```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+
+    :param str token: the JWT value for the currently propagated authentication
+        context
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, token=None, secrets=None):
+        self.token = token
+        self.secrets = secrets
+        self._payload = None
+        self._valid = None
+
+    def _secret(self):
+        if not self.secrets:
+            raise UndefinedSecretsException
+
+        return self.secrets.get_simple("jwt/authentication/secret")
+
+    def attach_context(self, context):
+        """Attach this authentication wrapper to the provided context
+
+        :param context: request context to attach this authentication to
+
+        """
+        context.authentication = self
+
+    @property
+    def valid(self):
+        """Validity of the current authentication context token
+
+        :type: bool
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and validated)
+
+        """
+        if self._valid is not None:
+            return self._valid
+        elif self.token is None:
+            return None
+
+        try:
+            self._payload = jwt.decode(self.token, self._secret(),
+                                       algorithms='RS256')
+            self._valid = True
+        except jwt.ExpiredSignatureError:
+            self._valid = False
+        except jwt.DecodeError:
+            self._valid = False
+
+        return self._valid
+
+    @property
+    def payload(self):
+        """Decrypted payload of the authentication token.
+
+        :type: dict
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+
+        """
+        if not self.valid:
+            return {}
+
+        return self._payload
+
+    @property
+    def account_id(self):
+        """Authenticated account_id for the current authenticated context
+
+        :type: account_id string or None if context authentication is invalid
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and the user_id returned)
+        :raises: :py:class:`UndefinedAuthenticationError` if there was no
+            authentication token defined for the current context
+
+        """
+        if self.token is None:
+            raise UndefinedAuthenticationError
+
+        return self.payload.get("sub", None)
```

Comment:
The `self.payload` property calls it, so it will raise it there.  If we want to not expose the payload itself, then the check would move back into here.